src/backend.py

```diff
@@ -0,0 +1,726 @@
+
+    # Create inner and outer pads
+    o_key_pad = bytes([b ^ 0x5c for b in key])
+    i_key_pad = bytes([b ^ 0x36 for b in key])
+
+    # Inner hash
+    h_inner = hashlib_sha256()
+    h_inner.update(i_key_pad + message)
+    inner_hash = h_inner.digest()
+
+    # Outer hash
+    h_outer = hashlib_sha256()
+    h_outer.update(o_key_pad + inner_hash)
+    return h_outer.digest()
+
+def require_auth(handler):
+    '''Decorator to require authentication using HMAC-SHA256'''
+    def auth_wrapper(request, *args, **kwargs):
+        global current_challenge, challenge_timestamp
+
+        if not current_challenge or not challenge_timestamp:
+            msg = json_dumps({"error": "Challenge missing"}), 401, 'application/json'
+            print(msg)
+            return msg
+
+        if (time() - challenge_timestamp) > CHALLENGE_EXPIRATION_SECONDS:
+            msg = json_dumps({"error": "Challenge expired"}), 401, 'application/json'
+            print(msg)
+            return msg
+
+        # Get the HMAC from the request headers
+        client_hmac = request.headers.get('X-Auth-HMAC') or request.headers.get('x-auth-hmac')
+        if not client_hmac:
+            msg = json_dumps({"error": "Missing authentication HMAC"}), 401, 'application/json'
+            print(msg)
+            print(request.headers)
+            return msg
+
+        # Retrieve stored password (the key)
+        credentials = ds_read_record("configuration", 0)
+        stored_password = credentials["Gpassword"].encode('utf-8')
+
+        # Construct the message string
+        path = request.path
+        query_string = request.query_string or ""
+        body_str = request.raw_data or ""
+        message_str = current_challenge + path + query_string + body_str
+        message_bytes = message_str.encode('utf-8')
+
+        # Compute expected HMAC
+        expected_digest = hmac_sha256(stored_password, message_bytes)
+        expected_hmac = hexlify(expected_digest).decode()
+
+        if client_hmac == expected_hmac:
+            # Authentication successful
+            current_challenge = None
+            challenge_timestamp = None
+            print("Authentication successful")
+            return handler(request, *args, **kwargs)
+        else:
+            # Authentication failed
+            print("Authentication failed")
+            print(f"Expected HMAC: {expected_hmac}")
+            print(f"Client HMAC: {client_hmac}")
+            print(message_bytes)
+            return json_dumps({"error": "Authentication failed"}), 401, 'application/json'
+
+    return auth_wrapper
+
+
+
+@add_route("/api/auth/challenge")
+def get_challenge(request):
+    global current_challenge, challenge_timestamp
+    # Generate a random nonce (challenge)
+    current_challenge = random_hex(64)
+    challenge_timestamp = time()
+    # Return the nonce to the client
+    return json_dumps({"challenge": current_challenge}), 200, 'application/json'
```

Comment:
The "challange" is what the client must hash with their copy of the password in order to prove they know it.